### PR TITLE
[CodeWhisperer] Add telemetrySending on users toggling cwspr autoSuggestion

### PR DESCRIPTION
**Pull request by Will-ShaoHua**
## Problem
CodeWhisperer team wants to track on how often time users make use of our auto-suggestion and make corresponding improvement from the metrics we collect.

Note: this PR is dependent on https://github.com/aws/aws-toolkit-vscode/pull/2831

## Solution
Send `aws_modifySetting` telemetry when user toggle cwspr auto-suggesion

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/vscode/commands2.ts [hunk 1]
```diff
@@ -45,7 +45,7 @@ export interface Command<T extends Callback = Callback> {
     execute(...parameters: Parameters<T>): Promise<ReturnType<T> | undefined>
 }
 
-interface RegisteredCommand<T extends Callback = Callback> extends Command<T> {
+export interface RegisteredCommand<T extends Callback = Callback> extends Command<T> {
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/vscode/commands2.ts:48]: Oh that's my bad. Forgot to change the return type.

File: src/test/testUtil.ts [hunk 2]
```diff
@@ -287,3 +288,14 @@ export async function closeAllEditors(): Promise<void> {
         throw new Error(`The following editors were still open after closeAllEditors():\n${editors.join('\n')}`)
     }
 }
+
+export function testCommand<T extends (...args: any[]) => unknown, U extends any[]>(
+    command: ReturnType<typeof Commands.declare<T, U>>,
+    ...args: U
+): Command<T> {
```
Comment (Will-ShaoHua) [inline src/test/testUtil.ts:295]:
```suggestion
): RegisteredCommand<T> {
```
@JadenSimon you mean we could change this right

File: src/shared/vscode/commands2.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Will-ShaoHua) [reply src/shared/vscode/commands2.ts:48]: got it

**Comment by Will-ShaoHua** [comment]
@JadenSimon @zixlin7 I will make this PR draft for now and wait for https://github.com/aws/aws-toolkit-vscode/pull/2831 being merged and update this PR again, thanks!

File: src/codewhisperer/commands/basicCommands.ts [hunk 3]
```diff
@@ -64,16 +73,27 @@ export const requestAccess = Commands.declare('aws.codeWhisperer.requestAccess',
 export const requestAccessCloud9 = Commands.declare(
     'aws.codeWhisperer.requestAccessCloud9',
     (context: ExtContext) => async () => {
```
Comment (JadenSimon) [inline src/codewhisperer/commands/basicCommands.ts:75]: Looks like we can make it even simpler which means simpler tests too:
```suggestion
    (globalState: vscode.Memento) => async () => {
```

Then change `context.extensionContext.globalState` -> `globalState` 
Same applies to the other command

File: src/codewhisperer/models/constants.ts [hunk 4]
```diff
@@ -18,6 +18,11 @@ export const CodeWhispererConstants = {
     invocationKeyThreshold: 15,
     specialCharactersList: ['{', '[', '(', ':', '\t', '\n'],
     normalTextChangeRegex: /[A-Za-z0-9]/g,
+    AutoSuggestion: {
+        settingId: 'cwspr_autoSuggestionActivation',
```
Comment (JadenSimon) [inline src/codewhisperer/models/constants.ts:22]: I know it's verbose but using the exact same 'namespace' with telemetry will help in the long run.
```suggestion
        settingId: 'codewhisperer_autoSuggestionActivation',
```

File: src/test/codewhisperer/commands/basicCommands.test.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+import { AwsContext } from '../../../shared/awsContext'
+import { SamCliContext } from '../../../shared/sam/cli/samCliContext'
+import { RegionProvider } from '../../../shared/regions/regionProvider'
+import { TelemetryService } from '../../../shared/telemetry/telemetryService'
+import { CredentialsStore } from '../../../credentials/credentialsStore'
+import { testCommand } from '../../shared/vscode/testUtils'
+import { Command } from '../../../shared/vscode/commands2'
+
+describe('CodeWhisperer-basicCommands', function () {
+    describe('CodeWhisperer-basicCommands', function () {
```
Comment (JadenSimon) [inline src/test/codewhisperer/commands/basicCommands.test.ts:25]: Duplicated?

File: src/test/codewhisperer/commands/basicCommands.test.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        let targetCommand: Command<any> & vscode.Disposable
+
+        beforeEach(function () {
+            resetCodeWhispererGlobalVariables()
+        })
+
+        afterEach(function () {
+            if (targetCommand) {
+                targetCommand.dispose()
+            }
```
Comment (JadenSimon) [inline src/test/codewhisperer/commands/basicCommands.test.ts:35]: nit:
```suggestion
            targetCommand?.dispose()
```

File: src/test/codewhisperer/commands/basicCommands.test.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        it('test set()', async function () {
+            const fakeMemeto = new FakeMemento()
+            set(CodeWhispererConstants.autoTriggerEnabledKey, true, fakeMemeto)
+            assert.strictEqual(fakeMemeto.get(CodeWhispererConstants.autoTriggerEnabledKey), true)
+
+            set(CodeWhispererConstants.autoTriggerEnabledKey, false, fakeMemeto)
+            assert.strictEqual(fakeMemeto.get(CodeWhispererConstants.autoTriggerEnabledKey), false)
+        })
+
+        it('test toggleCodeSuggestions: should emit aws_modifySetting event on user toggling autoSuggestion -- activate', async function () {
```
Comment (JadenSimon) [inline src/test/codewhisperer/commands/basicCommands.test.ts:64]: You had the right idea! This is a common pattern with the testing framework we use (`mocha`) and the convention is to make another describe block:
```suggestion
        describe('toggleCodeSuggestions', function () {
            it('should emit aws_modifySetting event on user toggling autoSuggestion -- activate', async function () {
```

File: src/test/codewhisperer/commands/basicCommands.test.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            const res = fakeExtContext.extensionContext.globalState.get(CodeWhispererConstants.autoTriggerEnabledKey)
+            assert.strictEqual(res, true)
+            assertTelemetryCurried('aws_modifySetting')({
+                settingId: CodeWhispererConstants.AutoSuggestion.settingId,
+                settingState: CodeWhispererConstants.AutoSuggestion.activated,
+            })
+        })
+
+        it('test toggleCodeSuggestions: should emit aws_modifySetting event on user toggling autoSuggestion - deactivate', async function () {
+            const fakeExtContext: ExtContext = {
```
Comment (JadenSimon) [inline src/test/codewhisperer/commands/basicCommands.test.ts:93]: Once we change the command input to `vscode.Memento` we can use `FakeMemento` instead.

**Review by JadenSimon — COMMENTED** [review]
Looks good to me after addressing comments. I appreciate the tests for `get`/`set` as well.

File: src/codewhisperer/models/constants.ts (hunk 4 above)
Comment (Will-ShaoHua) [reply src/codewhisperer/models/constants.ts:22]: make sense, will update it!

File: src/test/codewhisperer/commands/basicCommands.test.ts (hunk 5 above)
Comment (Will-ShaoHua) [reply src/test/codewhisperer/commands/basicCommands.test.ts:25]: no lol, I saw there are some other tests with this format, so...

Comment (Will-ShaoHua) [reply src/test/codewhisperer/commands/basicCommands.test.ts:25]: let me remove it

Comment (JadenSimon) [reply src/test/codewhisperer/commands/basicCommands.test.ts:25]: > no lol, I saw there are some other tests with this format, so...

Those probably slipped through code review :)

File: src/test/codewhisperer/commands/basicCommands.test.ts (hunk 6 above)
Comment (Will-ShaoHua) [reply src/test/codewhisperer/commands/basicCommands.test.ts:35]: ack!

File: src/test/codewhisperer/commands/basicCommands.test.ts (hunk 8 above)
Comment (Will-ShaoHua) [reply src/test/codewhisperer/commands/basicCommands.test.ts:93]: yea

**Review by JadenSimon — APPROVED** [review]

**Comment by JadenSimon** [comment]
Looks like there's a merge conflict, maybe just bump the telemetry version?

**Pull request merged**